### ISSUE TITLE
Upgrade from version 2.4.3 to 2.5 error

### ISSUE THREAD

**Issue by JavierNR** [thingsboard/thingsboard#2767]
Hello!
I have 2.4.3 CE version installation and tried to upgrade to 2.5 but running upgrade.sh throws an error:
```
Upgrading ThingsBoard from version 2.4.3 to 2.5 ...


Check the current PostgreSQL version...
PostgreSQL version is valid!
Load upgrade functions ...
Functions successfully loaded!
Updating timeseries schema ...

```

```
Failed to execute query: call create_partition_ts_kv_table() due to: ERROR: no existe la restricción «ts_kv_pkey» para la tabla «ts_kv_old»
  Where: sentencia SQL: «ALTER TABLE ts_kv_old
      RENAME CONSTRAINT ts_kv_pkey TO ts_kv_pkey_old»
función PL/pgSQL create_partition_ts_kv_table() en la línea 6 en sentencia SQL
Unexpected error during ThingsBoard installation!
java.lang.RuntimeException: Failed to execute query:call create_partition_ts_kv_table() due to:
        at org.thingsboard.server.service.install.AbstractSqlTsDatabaseUpgradeService.executeQuery(AbstractSqlTsDatabaseUpgradeService.java:113)
        at org.thingsboard.server.service.install.PsqlTsDatabaseUpgradeService.upgradeDatabase(PsqlTsDatabaseUpgradeService.java:93)
        at org.thingsboard.server.install.ThingsboardInstallService.performInstall(ThingsboardInstallService.java:139)
        at org.thingsboard.server.ThingsboardInstallApplication.main(ThingsboardInstallApplication.java:43)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.springframework.boot.loader.MainMethodRunner.run(MainMethodRunner.java:48)
        at org.springframework.boot.loader.Launcher.launch(Launcher.java:87)
        at org.springframework.boot.loader.Launcher.launch(Launcher.java:51)
        at org.springframework.boot.loader.PropertiesLauncher.main(PropertiesLauncher.java:597)
Caused by: org.postgresql.util.PSQLException: ERROR: no existe la restricción «ts_kv_pkey» para la tabla «ts_kv_old»
  Where: sentencia SQL: «ALTER TABLE ts_kv_old
      RENAME CONSTRAINT ts_kv_pkey TO ts_kv_pkey_old»
función PL/pgSQL create_partition_ts_kv_table() en la línea 6 en sentencia SQL
        at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2455)
        at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2155)
        at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:288)
        at org.postgresql.jdbc.PgStatement.executeInternal(PgStatement.java:430)
        at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:356)
        at org.postgresql.jdbc.PgStatement.executeWithFlags(PgStatement.java:303)
        at org.postgresql.jdbc.PgStatement.executeCachedSql(PgStatement.java:289)
        at org.postgresql.jdbc.PgStatement.executeWithFlags(PgStatement.java:266)
        at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:262)
        at org.thingsboard.server.service.install.AbstractSqlTsDatabaseUpgradeService.executeQuery(AbstractSqlTsDatabaseUpgradeService.java:99)
        ... 11 common frames omitted
Unexpected error during ThingsBoard installation!
ThingsBoard upgrade failed!
```

The machine had Postgres 9.6 installed and before upgrading to TB2.5 I've followed the instructions to upgrade Postgres to v11 and his cluster.

**Comment by ShvaykaD** [comment]
Hi @JavierNR , 

Please let me know if you created the database backup before the upgrade execution? If so, the best option to check what was going wrong during the upgrade script execution is to check your ThingsBoard database schema from the backup. 
Do you have the ability to restore the backup, or provide for us the backup to investigate the issue?
Probably, during the upgrading Postgres cluster the constraint was missed.

**Comment by JavierNR** [comment]
Hello @ShvaykaD,

Yes, I've created a database backup with `sudo -Hiu postgres pg_dumpall > mybackup.sql` the resulting file weights about 1.8 GB after that, I have followed the instructions in following order:

1. Get repo key and add to apt list
2. apt update and install postgresql-11
3. pg_dropcluster 11 main --stop
4. Stop thingsboard service
5. pg_upgradecluster 9.6 main -> Process went without problems
6. pg_dropcluster 9.6 main
7. Check if in postgresql.conf if cluster_name was set to 11/main
8. Install Thingsboard 2.5 release
9. Do upgrade.sh --fromVersion=2.4.3

I don't know the exact steps to restore the database but I'll try to figure out

**Comment by ShvaykaD** [comment]
Hi @JavierNR,

At the current stage, we can also check your ThingsBoard database schema. Probably the issue occurred only due to the missed ts_kv_pkey constraint.

in order to do this you need to login into your psql terminal:
psql -U postgres -d postgres -h 127.0.0.1 -W
Password: 
psql (11.7 (Ubuntu 11.7-2.pgdg18.04+1))
SSL connection (protocol: TLSv1.3, cipher: TLS_AES_256_GCM_SHA384, bits: 256, compression: off)
Type "help" for help.

connect to the ThingsBoard database:


postgres=# \c thingsboard
Password for user postgres: 
SSL connection (protocol: TLSv1.3, cipher: TLS_AES_256_GCM_SHA384, bits: 256, compression: off)
You are now connected to database "thingsboard" as user "postgres".

 and then execute the next command:

thingsboard=# \dt


The above command should show the database schema. Please add the output here.


In addition, I need to see also the schema of the next tables:

ts_kv (if it is exists),
ts_kv_old,

In order to do this you need to execute the next commands:

thingsboard=# \d+ ts_kv

thingsboard=# \d+ ts_kv_old


And the last one verification that needed is to execute the following select command:

thingsboard=# SELECT * FROM tb_schema_settings;


In order to help you, I need to see the outputs. Please adds them here.

**Comment by JavierNR** [comment]
Yes, I was doing this while you replied. I'm adding the data here:

\dt
 ```
Esquema |        Nombre        | Tipo  |     Dueño
---------+----------------------+-------+----------------
 public  | admin_settings       | tabla | thingsboard
 public  | alarm                | tabla | thingsboard
 public  | asset                | tabla | thingsboard
 public  | attribute_kv         | tabla | thingsboard
 public  | audit_log            | tabla | thingsboard
 public  | component_descriptor | tabla | thingsboard
 public  | customer             | tabla | thingsboard
 public  | dashboard            | tabla | thingsboard
 public  | device               | tabla | thingsboard
 public  | device_credentials   | tabla | thingsboard
 public  | entity_view          | tabla | thingsboard
 public  | event                | tabla | thingsboard
 public  | relation             | tabla | thingsboard
 public  | rule_chain           | tabla | thingsboard
 public  | rule_node            | tabla | thingsboard
 public  | tb_schema_settings   | tabla | thingsboard
 public  | tb_user              | tabla | thingsboard
 public  | tenant               | tabla | thingsboard
 public  | ts_kv                | tabla | thingsboard
 public  | ts_kv_latest         | tabla | thingsboard
 public  | user_credentials     | tabla | thingsboard
 public  | widget_type          | tabla | thingsboard
 public  | widgets_bundle       | tabla | thingsboard
(23 filas)
```

\d+ ts_kv

```
  Tabla «public.ts_kv»
   Columna   |            Tipo             | Ordenamiento | Nulable  | Por omisión | Almacenamiento | Estadísticas | Descripción
-------------+-----------------------------+--------------+----------+-------------+----------------+--------------+-------------
 entity_type | character varying(255)      |              | not null |             | extended       |              |
 entity_id   | character varying(31)       |              | not null |             | extended       |              |
 key         | character varying(255)      |              | not null |             | extended       |              |
 ts          | bigint                      |              | not null |             | plain          |              |
 bool_v      | boolean                     |              |          |             | plain          |              |
 str_v       | character varying(10000000) |              |          |             | extended       |              |
 long_v      | bigint                      |              |          |             | plain          |              |
 dbl_v       | double precision            |              |          |             | plain          |              |
Índices:
    "ts_kv_unq_key" UNIQUE CONSTRAINT, btree (entity_type, entity_id, key, ts)
```

ts_kv_old doesn't exist, probably the database wasn't modified with upgrade script

SELECT * FROM tb_schema_settings;

```
 schema_version
----------------
        2004003
```

**Comment by ShvaykaD** [comment]
@JavierNR Thank you.

We defined the reason for the issue, and it is occurred due to the bug in the upgrade script from version 2.4.2 to 2.4.3.

The constraints:

ts_kv_unq_key
ts_kv_latest_unq_key


doesn't exists in the 2.4.3 version, so in order to fix this please execute the next commands in your psql terminal:

postgres=# \c thingsboard

ALTER TABLE ts_kv DROP CONSTRAINT IF EXISTS ts_kv_pkey;
ALTER TABLE ts_kv DROP CONSTRAINT IF EXISTS ts_kv_unq_key;
ALTER TABLE ts_kv ADD CONSTRAINT ts_kv_pkey PRIMARY KEY (entity_type, entity_id, key, ts);


ALTER TABLE ts_kv_latest DROP CONSTRAINT IF EXISTS ts_kv_latest_pkey;
ALTER TABLE ts_kv_latest DROP CONSTRAINT IF EXISTS ts_kv_latest_unq_key;
ALTER TABLE ts_kv_latest ADD CONSTRAINT ts_kv_latest_pkey PRIMARY KEY (entity_type, entity_id, key);

After that please check that your ts_kv and ts_kv_latest schema has the valid constraints:

ts_kv_pkey;
 ts_kv_latest_pkey;

In order to do this execute the next commands:

/d+ ts_kv
/d+ ts_kv_latest

After that, if constrains is valid, execute the upgrade script again.

**Comment by ShvaykaD** [comment]
@JavierNR could you please, let me know if this helps you, in order to close this issue.

**Comment by JavierNR** [comment]
Hello,

As soon as I execute upgrade script, I see that it creates partition tables, and add keys to the dictionary but I'm getting this error:

`Failed to execute query: call insert_into_ts_kv('/tmp/ts_kv3493228088178935200/ts_kv.sql') due to: ERROR: debe ser superusuario o miembro del rol pg                                   _write_server_files para hacer COPY a un archivo`

I think it's just a permisions problem, so I'm going to GRANT the user that manages database to pg_write_server_files and let you know the results.

**Comment by JavierNR** [comment]
Now I have another error probably because script interrumpted previously and leave database with half tables or relations created

```
Updating timeseries schema ...
Failed to execute query: call create_partition_ts_kv_table() due to: ERROR: la relación «ts_kv_old» ya existe
  Where: sentencia SQL: «ALTER TABLE ts_kv
    RENAME TO ts_kv_old»
función PL/pgSQL create_partition_ts_kv_table() en la línea 4 en sentencia SQL
Unexpected error during ThingsBoard installation!
java.lang.RuntimeException: Failed to execute query:call create_partition_ts_kv_table() due to:
        at org.thingsboard.server.service.install.AbstractSqlTsDatabaseUpgradeService.executeQuery(AbstractSqlTsDatabaseUpgradeService.java:113)
        at org.thingsboard.server.service.install.PsqlTsDatabaseUpgradeService.upgradeDatabase(PsqlTsDatabaseUpgradeService.java:93)
        at org.thingsboard.server.install.ThingsboardInstallService.performInstall(ThingsboardInstallService.java:139)
        at org.thingsboard.server.ThingsboardInstallApplication.main(ThingsboardInstallApplication.java:43)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.springframework.boot.loader.MainMethodRunner.run(MainMethodRunner.java:48)
        at org.springframework.boot.loader.Launcher.launch(Launcher.java:87)
        at org.springframework.boot.loader.Launcher.launch(Launcher.java:51)
        at org.springframework.boot.loader.PropertiesLauncher.main(PropertiesLauncher.java:597)
Caused by: org.postgresql.util.PSQLException: ERROR: la relación «ts_kv_old» ya existe
  Where: sentencia SQL: «ALTER TABLE ts_kv
    RENAME TO ts_kv_old»
función PL/pgSQL create_partition_ts_kv_table() en la línea 4 en sentencia SQL
        at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2455)
        at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2155)
        at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:288)
        at org.postgresql.jdbc.PgStatement.executeInternal(PgStatement.java:430)
        at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:356)
        at org.postgresql.jdbc.PgStatement.executeWithFlags(PgStatement.java:303)
        at org.postgresql.jdbc.PgStatement.executeCachedSql(PgStatement.java:289)
        at org.postgresql.jdbc.PgStatement.executeWithFlags(PgStatement.java:266)
        at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:262)
        at org.thingsboard.server.service.install.AbstractSqlTsDatabaseUpgradeService.executeQuery(AbstractSqlTsDatabaseUpgradeService.java:99)
        ... 11 common frames omitted
Unexpected error during ThingsBoard installation!
```

**Comment by ShvaykaD** [comment]
@JavierNR Could you please specify which OS you are using?

**Comment by JavierNR** [comment]
Yes, Debian 9.12

**Comment by ShvaykaD** [comment]
As far as I know, we don't have the upgrade scenarios for Debian distributions. Probably this is the reason for that issue because we didn't test the upgrade on Debian. 

In this case, please do the backup:

sudo -Hiu postgres pg_dump thingsboard > thingsboard.sql.bak

and then try to execute the next commands:

# Drop new ts_kv table. Please be sure that you will remove the new table without data.
DROP TABLE ts_kv;

# Drop ts_kv_dictionary table
DROP TABLE ts_kv_dictionary;

# rename the old table to ts_kv.
ALTER TABLE ts_kv_old RENAME to ts_kv;

# rename the old table constraint to the ts_kv_pkey
ALTER TABLE ts_kv RENAME CONSTRAINT ts_kv_pkey_old TO ts_kv_pkey;

and execute the script again.

Hope this will help you.

**Comment by ashvayka** [comment]
Hi @JavierNR @ShvaykaD 

I am afraid last comment from @ShvaykaD may miss lead some users that usually scroll this threads until they see the commands and simply execute them. This may cause troubles. The commands above are related only to restoring the database after failed attempt of upgrade. 

The actual solution to prevent the upgrade failure is now documented in the Upgrade Instructions on the website. I am adding it here. Please execute the following commands before running the upgrade script:

```sql
# Connect to thingsboard database:
\c thingsboard

# Execute the next commands:

# Update ts_kv table constraints:
ALTER TABLE ts_kv DROP CONSTRAINT IF EXISTS ts_kv_unq_key;
ALTER TABLE ts_kv DROP CONSTRAINT IF EXISTS ts_kv_pkey;
ALTER TABLE ts_kv ADD CONSTRAINT ts_kv_pkey PRIMARY KEY (entity_type, entity_id, key, ts);

# Update ts_kv_latest table constraints:
ALTER TABLE ts_kv_latest DROP CONSTRAINT IF EXISTS ts_kv_latest_unq_key;
ALTER TABLE ts_kv_latest DROP CONSTRAINT IF EXISTS ts_kv_latest_pkey;
ALTER TABLE ts_kv_latest ADD CONSTRAINT ts_kv_latest_pkey PRIMARY KEY (entity_type, entity_id, key);

# exit psql terminal 
\q
```

**Comment by IoTopenTech** [comment]
Dear @ashvayka 
I assume that this procedure is only applicable if you previously upgrades from 2.4.2 to 2.4.3.
I started with ThingsBoard 2.4.3, so I should't apply this procedure. am I right?
Besides, I have spotted a misspelling in the Ubuntu installation instructions
![imagen](https://user-images.githubusercontent.com/52624907/81898985-16f78f80-95ba-11ea-9572-f2ee0bf9c300.png)

**Comment by YevhenBondarenko** [comment]
Hi @IoTopenTech.

Thanks for the found mistake. We have changed kafka version in the instructions.

**Comment by JavierNR** [comment]
Hi @ShvaykaD!
First I've checked tables to be dropped have NO data, and checked ts_kv_old exists with old data. As @ashvayka says this can lead to people executing wrong commands and delete their telemetries.

After renaming and dropping the tables, upgrade script went perfectly and now installation is working.

Thank you for your great support and great software

**Comment by russelmahmud** [comment]
I am getting this error while migrating in Amazon RDS. 
```
Failed to execute query: call insert_into_ts_kv('/tmp/ts_kv2267233513583762561/ts_kv.sql') due to: ERROR: must be superuser or a member of the pg_write_server_files role to COPY to a file
Hint: Anyone can COPY to stdout or from stdin. psql's \copy command also works for anyone.
```
I couldn't find anyway to give permission `pg_write_server_files` for the master user of RDS instance. Has anyone faced the same problem?

**Comment by ashvayka** [comment]
This is fixed in 2.5.1 and the upgrade from 2.4.3 to 2.5.1 will work. See my comments here: https://github.com/thingsboard/thingsboard/issues/2880#issuecomment-636837629
